### ISSUE TITLE
Incorrect working of "track" library under R.NET

### ISSUE THREAD

**Issue by RebelionTheGrey** [jmp75/rdotnet#21]
Hello. 

I found one problem when tried to work with "track" package. I really can't understand the reason of it, but...
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using RDotNet;
using RDotNet.NativeLibrary;
using RDotNet.Dynamic;

namespace ConsoleApplication10
{
    class Program
    {
        static void Main(string[] args)
        {
            var engine = REngine.GetInstance();

            engine.Initialize();

            engine.Evaluate("library(track);");
            engine.Evaluate(@"track.start(clobber = ""vars"", dir = file.path(tempdir(), 'rdata12356'))");

            engine.Evaluate("a <- 5;");
            engine.Evaluate("print(track.summary());");
            engine.Evaluate("print(ls(globalenv()));");

            Console.ReadLine();

        }
    }
}
track.summary() doesn't work properly via R.NET. 
This is execution result via R.NET:
Tracking <env R_GlobalEnv> (writable) using existing directory 'rdatadir'
[1] class    mode     extent   length   size     modified TA       TW
<0 строк> (или 'row.names' нулевой длины)
[1] "a"
This is execution result via R-3.2.2.x64
> library(track);
> track.start(clobber = "vars", dir = file.path(tempdir(), 'rdata123'));
Tracking <env R_GlobalEnv> (writable) using new directory 'C:\Users\Rebelion\AppData\Local\Temp\RtmpGmyc4B/rdata123'
> 
> a <- 5;
> print(track.summary());
    class    mode extent length size            modified TA TW
a numeric numeric    [1]      1   48 2015-11-18 20:36:23  0  1
> print(ls(globalenv()));
[1] "a"
> 
> track.stop();
Stopping tracking on <env R_GlobalEnv>
What's the reason of this difference in working? 

Thank you.

**Comment by RebelionTheGrey** [comment]
So, if add code like this:
```c#
engine.Evaluate("print(track.status());");
```
we will see, that variables are untracked:
    Status InMem FileBase FileExists Saved
    a untracked    NA     <NA>         NA    NA

It's so strange. 
If I make some improvements like permanent path, or remove tempdir from
```C#
engine.Evaluate(@"track.start(clobber = ""vars"", dir = file.path(tempdir(), 'rdata12356'))");
```
(make something like this)
```C#
engine.Evaluate(@"track.start(auto = TRUE, pos = 1, clobber = ""vars"", dir = file.path('C:/User Root/2'));");
```
I'll get, that at first runvariables are untracked, but on the second run they are tracked.

I think problem in task callback functions which doesn't call correct in R.NET. Maybe I wrong.
(It's from "track" package)

    A tracking environment can be linked to one environment on the search path, but the tracking environment is not on the search path itself. An environment on the search path can only have one tracking environment linked to it. In standard use, variables are tracked automatically by a task callback function. Alternatively, variables to track can be registered with the tracking environment using the function track().

**Comment by skyguy94** [comment]
I'm not seeing a difference between R.Net and rstudio. If I run this code in RStudio
```
library(track);
track.start(clobber = "vars", dir = file.path(tempdir(), 'rdata12356'))
a <- 5;
print(track.summary());
print(ls(globalenv()));
```
I get this as a result on the first run ```
Tracking <env R_GlobalEnv> (writable) using new directory 'C:\Users\Ritch\AppData\Local\Temp\Rtmp8MGX25/rdata12356'
[1] class    mode     extent   length   size     modified TA       TW      
<0 rows> (or 0-length row.names)
[1] "a"
```

That's the exact same result I get from R.Net

**Comment by RebelionTheGrey** [comment]
It's really strange - I don't often use RStudio, I use Console for testing primary raw code. So, what's the difference maybe between R.NET + RStudio and R Console, because in R Console this code works well...

**Comment by RebelionTheGrey** [comment]
That's sooooo strange! As I told earlier, it could be connected with task callbacks.

Look at this code I've got from help:

    times <- function(total = 3, str = "Task a") {
                  ctr <- 0

                 function(expr, value, ok, visible) {
                                ctr <<- ctr + 1
                                cat(str, ctr, "\n")
                                if(ctr == total) {
                                                       cat("handler removing itself\n")
                                                      }
                                return(ctr < total)
                 }
     }
     
     # add the callback that will work for
     # 4 top-level tasks and then remove itself.
     n <- addTaskCallback(times(4))

     # now remove it, assuming it is still first in the list.
     removeTaskCallback(n)

     ## Not run: 
     # There is no point in running this
     # as
     addTaskCallback(times(4))

     sum(1:10)
     sum(1:10)
     sum(1:10)
     sum(1:10)
     sum(1:10)

     ## End(Not run)

If It was ran in R Terminal, we have this output:

    1 
    1 
    Task a 1 
    > 
    >  sum(1:10)
    [1] 55
    Task a 2 
    >  sum(1:10)
    [1] 55
    Task a 3 
    >  sum(1:10)
    [1] 55
    Task a 4 
    handler removing itself
    >  sum(1:10)
    [1] 55
    >  sum(1:10)
    [1] 55
    > 
    > ## End(Not run)


But, this code, was ran from RStudio, gave me this:

    > source('~/tsts2.R')
    Task a 1 
    > 

And this code, was ran in R.NET, gave me this:

    Multithreaded BLAS/LAPACK libraries detected. Using 2 cores for math algorithms.
    [1] TRUE
    1
    1
    [1] 55
    [1] 55
    [1] 55
    [1] 55
    [1] 55

Amazing! No callbacks! Why? Carl, why??? :-(

**Comment by RebelionTheGrey** [comment]
http://stackoverflow.com/questions/33887671/strange-addtaskcallback-work-in-rstudio/33908486#33908486  
RStudio also works normal, execute this code well if not using 'source(...)'.

**Comment by skyguy94** [comment]
Have you tried using source in R.Net as suggested in that SO link? RStudio and R.Net use the publicly documented API for invoking the R engine, so what works in one should work in the other.

The SO post you linked to explains one of the limitations in getting R to parse statements through this API, namely, "When you're executing code line-by-line, each statement executed will return control to the top level, and callbacks will execute."

**Comment by RebelionTheGrey** [comment]
skyguy94, sorry, it's not an attempt to blame jmp75 or your in something, it's just an attempt to understand of the root of problem - R limitation, R.NET realization or my own curved hands. :-( 
If I use example code for addTaskCallback in RStudio in line-by-line mode without using "source", I get complementary working of RStudio to R Terminal. Also it works for this code:

    library(track);
    track.start(clobber = "vars", dir = file.path('c:/User Root/2/'))
    a <- 5;
    print(track.summary());
    print(ls(globalenv()));

I have "a" tracked if I run code line-by-line in RStudio (it works like R Terminal), but execution this code with using "Evaluate" to each line of earlier example for "track" or using "Evaluate" for whole R code in example for "track" and also for addTaskCallback. 

Thank you and sorry you spended time on this. :-(

**Comment by skyguy94** [comment]
I don't think you are trying to blame or anything and I'm just trying to help, but there are some limitations with how R.Net sends the R code to the eval function. Generally, I test things out in RStudio as I'd like R.Net to have parity with how it does things. In this case, I think R.Net should be able to handle multiline statements properly, but it does not. At this point in time, I think the suggested source workaround is the best approach.
